Lean4Axiomatic/ClassicalAlgebra/Group.lean

```diff
@@ -0,0 +1,108 @@
+import Lean4Axiomatic.AbstractAlgebra.Substitutive
+import Lean4Axiomatic.ClassicalAlgebra.Monoid
+
+namespace Lean4Axiomatic.CA.Group
+
+open Relation.Equivalence (EqvOp)
+
+
+/-!
+A formalization of Group using multiplicative notation.
+-/
+
+/-! ### Definitions -/
+
+/--
+Operations for Group, namely the binary operation, identity element, and
+existence of inverses.
+-/
+class Ops (α : Type) :=
+  binop : α → α → α
+  ident : α
+  inverse : (x : α) → α
+
+export Ops (ident inverse)
+
+/-- Enables the use of the `· * ·` operator for f (the monoid operation). -/
+local instance group_mul_op_inst {α : Type} [Ops α] : Mul α := {
+  mul := Group.Ops.binop
+}
+
+/-- Properties of Group. -/
+class Props (α : Type) [EqvOp α] [Ops α] :=
+  substL {x y z : α} : x ≃ y → x * z ≃ y * z
+  substR {x y z : α} : x ≃ y → z * x ≃ z * y
+  assoc {x y z : α} : (x * y) * z ≃ x * (y * z)
+  identL {x : α} : ident * x ≃ x
+  identR {x : α} : x * ident ≃ x
+  inverse_propL (x : α) : (inverse x) * x ≃ ident
+  inverse_propR (x : α) : x * (inverse x) ≃ ident
+
+export Props (
+  substL substR assoc identL identR inverse_propL inverse_propR
+)
+
+/-- All axioms for generic types to form a Group. -/
+class Group (α : Type) [EqvOp α] :=
+  toOps : Group.Ops α
+  toProps : Group.Props α
+
+attribute [instance] Group.toOps
+attribute [instance] Group.toProps
+
+/-! ### Properties -/
+
+variable {α : Type} [EqvOp α] [g : Group α]
+
+/-- Enables the use of `AA.substL`, `AA.substR`, etc. -/
+instance group_subst_inst
```

Comment:
You should make this a `local instance`, to avoid introducing a conflicting instance value when defining a group on a type that already has a `Substitutive₂` instance for the binary operator (such as the natural numbers).